kore/src/Kore/Unification/Procedure.hs

```diff
@@ -64,8 +69,24 @@ unificationProcedure sideCondition p1 p2
     let (term, conditions) = Conditional.splitTerm pat
     orCeil <- makeEvaluateTermCeil sideCondition term
     ceil' <- Monad.Unify.scatter orCeil
-    lowerLogicT . simplifyCondition sideCondition
-        $ Conditional.andCondition ceil' conditions
+    condition <-
+        lowerLogicT . simplifyCondition sideCondition
+            $ Conditional.andCondition ceil' conditions
+    pure $ assertNoFreeEquationVariableName condition
   where
     p1Sort = termLikeSort p1
     p2Sort = termLikeSort p2
+
+    assertNoFreeEquationVariableName
+        :: Condition RewritingVariableName
+        -> Condition RewritingVariableName
+    assertNoFreeEquationVariableName (Conditional () predicate substitution) =
+        let
+            predicate' =
+                assert (not . freeEquationVariableName $ predicate) predicate
+            substitution' =
+                mapTerms
+                    (\t -> assert (not . freeEquationVariableName $ t) t)
+                    substitution
+        in
+            Conditional () predicate' substitution'
```

Comment:
This assertion should appear in `Kore.Equation.Application` after matching and unification on function arguments.